tools/bitfield_gen.md

```diff
@@ -0,0 +1,672 @@
+containing an array of the architecture word type (`uint32` or `uint64`) of the
+length computed from the number and width of fields in the block. Given that the
+memory layout is fully specified it is possible to operate on these raw values
+directly without breaking the semantics of the generated functions.
+
+This is usually neither recommended nor necessary, and for code undergoing
+formal verification strongly discouraged, because it introduces additional
+verification overhead.
+
+However, one special case is occasionally convenient: when initialising an area
+of memory with 0, and reinterpreting that memory as a bitfield structure, we are
+guaranteed to get a value where all fields are 0. Similarly, returning an all-0
+struct from a function instead of calling the constructor with only 0 parameters
+may be acceptable, although one should check the generated assembly if the
+compiler was not able to optimise the constructor away -- if it does then the
+constructor form is clearer and easier to use in proofs.
+
+The tool will not access padding fields, but is not guaranteed to preserve them
+when copying. Writing/reading padding fields is not guaranteed to be stable.
+
+### Architecture Parameters and Canonical Pointers
+
+The tool supports architectures with 32 and 64 bit word sizes. The command
+
+```bf_gen
+base 32
+```
+
+selects a 32 bit native word size.
+
+Many 64 bit architectures have the concept of canonical pointers where not all
+64 bits of the available space are usable for pointers, but only for example 48
+bits. In this setting a pointer value would canonical if and only if all top
+bits are set when bit 47 (counting from 0) is set. This is equivalent to sign
+extension, i.e. treating bit 47 as the sign bit and extend that sign bit up to
+bit 63.
+
+A potential use for this is for kernel pointers to always have bit 47 set (and
+all bits above) and for user pointers to always have bit 47 unset (and all bits
+above).
+
+The bitfield generator allows specifying this canonical bit and whether the top
+bits for pointers should be sign extended or not like this:
+
+```bf_gen
+base 64(48,1)
+```
+
+This means: word size is 64, pointers are 48 bits `[0..47]`, and bit 48 and
+above should be set to 1 if bit 47 is set.
+
+This only affects `field_high` fields which are assumed to be pointers.
+
+```bf_gen
+base 64(48,0)
+```
+
+means word size is 64, pointers are at most 48 bits, and no sign extension. This
+is currently equivalent to just `base 64`.
+
+## Syntax Reference
+
+### Lexical Structure
+
+The generator distinguishes identifiers, integer literals, comments, white
+space, keywords, and operators. Keywords and operators are shown inline in the
+grammar in the next section. The rest are define below.
+
+A multi-letter ident must start with a letter or underscore, followed by
+letters, underscores or numbers. A one-letter ident must consist of one
+letter. Letters are Latin letters `[a-z]` and `[A-Z]`, digits are `[0-9]`.
```

Comment:
I am assuming that this is what you meant here. If it is, then I think it would be nicer to distinguish letters and characters.
```suggestion
A multi-character ident must start with a letter or underscore, followed by
letters, underscores or numbers. A one-character ident must consist of one
letter. Letters are Latin letters `[a-z]` and `[A-Z]`, digits are `[0-9]`.
```

Reply:
You mean because `_` and `[0-9]` are not letters? Yes, I think your change makes sense.